elastic-apm-agent/src/main/java/co/elastic/apm/agent/premain/ShadedClassLoader.java

```diff
@@ -204,11 +228,26 @@ private byte[] getShadedClassBytes(String name) throws ClassNotFoundException {
      * if the resource could not be found, or if the loader is closed.
      */
     @Override
-    public URL findResource(String name) {
-        if (locallyNonAvailableResources.get().contains(name)) {
-            return null;
+    public URL findResource(final String name) {
+        URL result = null;
+        if (!locallyNonAvailableResources.get().contains(name)) {
+
+            // while most of the body of default 'findResource' in JDK implementation is in a privileged action
+            // an extra URL check is performed just after it, hence we have to wrap the whole method call in a privileged
+            // action otherwise the security manager will complain about lack of proper read privileges on the agent jar
+            result = AccessController.doPrivileged(new PrivilegedAction<URL>() {
+                @Override
+                public URL run() {
+                    return findResourceInternal(getShadedResourceName(name));
+                }
+            });
```

Comment:
[for reviewer] this fixes the `NoClassDefFound` exceptions, the JDK implementation of `findResource` that we call through the `super.findResource(name)` performs an extra URL check outside of a privileged action.